File: norvig/discrete.R
library(MASS)

pmf <- function(X) Y <- {
  if (is.vector(X)) {
    df <- as.data.frame(table(X) / length(X))
  } else if (is.data.frame(X)) {
    stopifnot(ncol(X) == 2)
    df <- X
  }
  
  colnames(df) <- c('value', 'prob')
  total <- sum(df$prob)
  df$prob <- df$prob/total
  
  df
}

such_that <- function(X, f) vapply(X, f, logical(1))

P <- function(event, space) {
  if (is.function(event)) {
    subset <- such_that(space, event)
  } else {
    subset <- event
  }

  if(is.table(space)) {
    prob <- sum(space[subset, ]$prob)
  } else {
    prob <- fractions(length(space[subset])/length(space))
  }
  
  prob
}

# Testing
# such_that(seq(0, 12), function(n) n %% 2 == 0)
# dane_data <- data.frame(pairs = c('GG', 'GB', 'BG', 'BB'), nos = c(121801, 126840, 127123, 135138))
# P2 <- pmf(dane_data)
# P1 <- pmf(sample.int(10, 20, TRUE))

File: norvig/discrete.py
from collections.abc import Iterable
from typing import Any, Callable, Dict, Set, TypeVar, Union
from fractions import Fraction

import itertools

U = TypeVar('U')
type Predicate = Callable[..., bool]


def cross(A: Iterable[Any], B: Iterable[Any]) -> Set[str]:
    """
    Returns the cross product of two iterable containers A and B in the form
    of a concatenated string.
    """
    return {str(a) + str(b) for a in A for b in B}


def all_combos(items: Iterable[str], r: int, sep: str = ' ') -> Set[str]:
    """
    Returns all combinations of r items from an iterable collection of n items
    as a set of strings. Each string is a list of items from the collection
    separated by sep.
    """

    return {sep.join(c) for c in itertools.combinations(items, r)}


class Pmf(Dict[Any, float]):
    """
    Probability mass function.
    """

    def __init__(self, mapping=(), **kwargs):
        self.update(mapping, **kwargs)

        total = sum(self.values())
        for outcome in self:
            self[outcome] = self[outcome] / total
            assert self[outcome] >= 0


def such_that(f: Predicate, X: Union[Iterable, Pmf]) -> set:
    """
    Returns a subset of elements x the set X that satisfy the predicate.
    """
    if isinstance(X, Pmf):
        return Pmf({e: X[e] for e in X if f(e)})
    else:
        return {x for x in X if f(x)}


def P(event: Union[Predicate, Set[U]],
      space: Union[Set[U], Pmf]) -> Union[float, Fraction]:
    """
    Returns the probability of an event in a sample space.

    The event may be specified as a predicate or a set.    
    """
    if callable(event):
        subset = such_that(event, space)
    else:
        subset = event

    if isinstance(space, Pmf):
        return sum(space[x] for x in space if x in subset)
    else:
        return Fraction(len(subset & space), len(space))


def joint_pmf(P1: Pmf, P2: Pmf, sep: str = '') -> Pmf:
    """
    Returns the joint probability distribution of two probability mass functions.

    The events in the joint distribution are strings formed of concatenation of
    events of P1 and P2 separated by sep.
    """
    return Pmf({e1 + sep + e2: P1[e1] * P2[e2] for e1 in P1 for e2 in P2})
